index.js

```diff
@@ -301,51 +317,62 @@ const execa = (command, args, options) => {
 	}
 
 	// TODO: Use native "finally" syntax when targeting Node.js 10
-	const handlePromise = () => pFinally(Promise.all([
-		processDone,
-		getStream(spawned, 'stdout', {encoding, buffer, maxBuffer}),
-		getStream(spawned, 'stderr', {encoding, buffer, maxBuffer}),
-		getStream(spawned, 'all', {encoding, buffer, maxBuffer: maxBuffer * 2})
-	]).then(results => { // eslint-disable-line promise/prefer-await-to-then
-		const result = results[0];
-		result.stdout = results[1];
-		result.stderr = results[2];
-		result.all = results[3];
-
-		if (result.error || result.code !== 0 || result.signal !== null || isCanceled) {
-			const error = makeError(result, {
-				joinedCommand,
-				parsed,
-				timedOut,
-				isCanceled
-			});
-
-			// TODO: missing some timeout logic for killed
-			// https://github.com/nodejs/node/blob/master/lib/child_process.js#L203
-			// error.killed = spawned.killed || killed;
-			error.killed = error.killed || spawned.killed;
+	const handlePromise = () => {
+		let processComplete = Promise.all([
+			processDone,
+			getStream(spawned, 'stdout', {encoding, buffer, maxBuffer}),
+			getStream(spawned, 'stderr', {encoding, buffer, maxBuffer}),
+			getStream(spawned, 'all', {encoding, buffer, maxBuffer: maxBuffer * 2})
+		]);
+
+		if (timeoutId) {
+			processComplete = Promise.race([
+				processComplete,
+				resolvable
+			]);
+		}
 
-			if (!parsed.options.reject) {
-				return error;
+		return pFinally(processComplete.then(results => { // eslint-disable-line promise/prefer-await-to-then
```

Comment:
I would prefer using async/await here. If ESLint complains, it's usually a good reason for it. You could extract everything inside the `.then()` here into an async function that awaits `processComplete` first. Then you could just call it here: `pFinally(finalize(), destroy)`

Reply:
FWIW, I didn't write this code, just moved it around. Your suggestion is clearer, so I've implemented it.